exotica/src/Scene.cpp

```diff
@@ -840,9 +845,87 @@ namespace exotica
   void Scene::cleanScene()
   {
       getPlanningScene()->removeAllCollisionObjects();
+      updateSceneFrames();
       collision_scene_->reinitializeCollisionWorld();
   }
 
+  void Scene::updateSceneFrames()
+  {
+      kinematica_.resetModel();
+
+      planning_scene::PlanningScenePtr ps = collision_scene_->getPlanningScene();
+
+      // Add world objects
+      for(auto& object : *ps->getWorld())
+      {
+          if(object.second->shapes_.size())
+          {
+              // Use the first collision shape as the origin of the object
+              Eigen::Affine3d objTransform = object.second->shape_poses_[0];
+              std::string objectName = object.first;
+              kinematica_.AddElement(objectName, objTransform);
+              for(int i=0; i<object.second->shape_poses_.size();i++)
+              {
+                  Eigen::Affine3d trans = objTransform.inverse()*object.second->shape_poses_[i];
+                  kinematica_.AddElement(objectName+"_collision_"+std::to_string(i), trans, objectName, object.second->shapes_[i]);
+              }
+          }
+          else
+          {
+            HIGHLIGHT("Object with no shapes ('"<<object.first<<"')");
+          }
+      }
+
+      // Add robot collision objects
+      ps->getCurrentStateNonConst().update(true);
+      const std::vector<const robot_model::LinkModel*>& links =
+          ps->getCollisionRobot()->getRobotModel()->getLinkModelsWithCollisionGeometry();
+      for (int i = 0; i < links.size(); ++i)
+      {
+          Eigen::Affine3d objTransform = ps->getCurrentState().getGlobalLinkTransform(links[i]);
+
+          for (int j = 0; j < links[i]->getShapes().size(); ++j)
+          {
+              Eigen::Affine3d trans = objTransform.inverse()*ps->getCurrentState().getCollisionBodyTransform(links[i], j);
+              kinematica_.AddElement(links[i]->getName()+"_collision_"+std::to_string(j), trans, links[i]->getName(), links[i]->getShapes()[j]);
+          }
+      }
+
+      kinematica_.UpdateModel();
+  }
+
+  void Scene::attachObject(const std::string& name, const std::string& parent)
+  {
+      kinematica_.changeParent(name, parent, KDL::Frame(), false);
+      attached_objects_[name] = AttachedObject(parent);
+  }
+
+  void Scene::attachObject(const std::string& name, const std::string& parent, const KDL::Frame& pose)
+  {
+      kinematica_.changeParent(name, parent, pose, true);
+      attached_objects_[name] = AttachedObject(parent, pose);
+  }
+
+  void Scene::detachObject(const std::string& name)
+  {
+      if(!hasAttachedObject(name)) throw_pretty("'"<<name<<"' is not attached to the robot!");
+      auto object = attached_objects_.find(name);
+      kinematica_.changeParent(name, "", KDL::Frame(), false);
+      attached_objects_.erase(object);
+  }
+
+  bool Scene::hasAttachedObject(const std::string& name)
+  {
+      if(attached_objects_.find(name)!=attached_objects_.end())
```

Comment:
Can't this be reduced to ``return (attached_objects_.find(name)!=attached_objects_.end()`` or ``return (attached_objects_.find(name)!=attached_objects_.end() ? true : false``